### PR TITLE
useInstanceId: Convert to typescript

### PR DESCRIPTION
**Pull request by walbo**
## What?
Refactors `useIntanceId` to TypeScript instead of relying on JSDocs.

## Why?
When using `useInstanceId` in a typescript file it always returns `string | number`. By refactor `useInstanceId` to typescript we can return a specific type based on what params you pass to the function.

## How?
Refactored `useIntanceId` to TypeScript.

### Before
```ts
// Passing only a object returns string or number
useInstanceId( {} ) => string | number

// Passing object and prefix returns string or number
useInstanceId( {}, 'prefix' ) => string | number

// Passing object, prefix and a string as preferred id returns string or number
useInstanceId( {}, 'prefix', 'preferredId' ) => string | number

// Passing object, prefix and a number as preferred id returns string or number
useInstanceId( {}, 'prefix', 2022 ) => string | number
```

### After
```ts
// Passing only a object returns a number
useInstanceId( {} ) => number

// Passing object and prefix returns a string
useInstanceId( {}, 'prefix' ) => string

// Passing object, prefix and a string as preferred id returns a string
useInstanceId( {}, 'prefix', 'preferredId' ) => string

// Passing object, prefix and a number as preferred id returns a number
useInstanceId( {}, 'prefix', 2022 ) => number
```

### CODE REVIEW

File: packages/compose/src/hooks/use-instance-id/index.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// Disable reason: Object and object are distinctly different types in TypeScript and we mean the lowercase object in thise case
+// but eslint wants to force us to use `Object`. See https://stackoverflow.com/questions/49464634/difference-between-object-and-object-in-typescript
+/* eslint-disable jsdoc/check-types */
```
Comment (ciampo) [inline packages/compose/src/hooks/use-instance-id/index.ts]: I believe we can remove this comment

File: packages/compose/src/hooks/use-instance-id/index.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
+const instanceMap = new WeakMap< object, number >();
+
+/**
+ * Creates a new id for a given object.
+ *
+ * @param {object} object Object reference to create an id for.
```
Comment (ciampo) [inline packages/compose/src/hooks/use-instance-id/index.ts]: Let's remove the types from all JSDoc comments — specifying types in TypeScript is enough

File: packages/compose/src/hooks/use-instance-id/index.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ * 2. When object and prefix is given, the returned value is a string
+ * 3. When preferredId is given, the returned value is the type of preferredId
+ */
+function useInstanceId( object: object ): number;
+function useInstanceId( object: object, prefix: string ): string;
+function useInstanceId< T extends string | number >(
+	object: object,
+	prefix: string,
+	preferredId?: T
+): T;
```
Comment (ciampo) [inline packages/compose/src/hooks/use-instance-id/index.ts:36]: Nice usage of TypeScript overloading!

**Review by ciampo — COMMENTED** [review]
Great work here!

I've left a few comments re. cleaning the code up, but I've already tested the changes and they seem to work fine 🚀 

The `compose` package also seems to have a [CHANGELOG](https://github.com/WordPress/gutenberg/blob/f5cd9e1285c17af5431b391fe752560af0a95dc2/packages/compose/CHANGELOG.md), we should probably add an entry there as well

File: packages/compose/src/hooks/use-instance-id/index.ts (hunk 1 above)
Comment (walbo) [reply packages/compose/src/hooks/use-instance-id/index.ts]: Yes, can be removed when we only specifies types in TypeScript 👍 

Fixed in [0aa7b42](https://github.com/WordPress/gutenberg/pull/43790/commits/0aa7b42b6dbe61b1c51439d73e1558f633d1cf92)

File: packages/compose/src/hooks/use-instance-id/index.ts (hunk 2 above)
Comment (walbo) [reply packages/compose/src/hooks/use-instance-id/index.ts]: 👍 Done in [0aa7b42](https://github.com/WordPress/gutenberg/pull/43790/commits/0aa7b42b6dbe61b1c51439d73e1558f633d1cf92)

**Review by ciampo — APPROVED** [review]
🚀

**Pull request merged**